packages/client/src/publish/MessageFactory.ts

```diff
@@ -0,0 +1,119 @@
+import { random } from 'lodash'
+import { EncryptedGroupKey, EncryptionType, EthereumAddress, StreamID, StreamMessage, toStreamPartID } from 'streamr-client-protocol'
+import { EncryptionUtil } from '../encryption/EncryptionUtil'
+import { GroupKeyId } from '../encryption/GroupKey'
+import { createRandomMsgChainId, MessageChain } from './MessageChain'
+import { MessageMetadata } from './Publisher'
+import { keyToArrayIndex } from '@streamr/utils'
+import { GroupKeySequence } from './GroupKeyQueue'
+import { Mapping } from '../utils/Mapping'
+
+export interface MessageFactoryOptions {
+    publisherId: EthereumAddress
+    streamId: StreamID
+    getPartitionCount: (streamId: StreamID) => Promise<number>
+    isPublicStream: (streamId: StreamID) => Promise<boolean>
+    isPublisher: (streamId: StreamID, publisherId: EthereumAddress) => Promise<boolean>
+    createSignature: (payload: string) => Promise<string>
+    useGroupKey: () => Promise<GroupKeySequence>
+}
+
+export class MessageFactory {
+
+    private readonly publisherId: EthereumAddress
+    private readonly streamId: StreamID
+    private defaultPartition: number | undefined
+    private readonly defaultMessageChainIds: Mapping<[partition: number], string>
+    private readonly messageChains: Mapping<[partition: number, msgChainId: string], MessageChain>
+    private readonly getPartitionCount: (streamId: StreamID) => Promise<number>
+    private readonly isPublicStream: (streamId: StreamID) => Promise<boolean>
+    private readonly isPublisher: (streamId: StreamID, publisherId: EthereumAddress) => Promise<boolean>
+    private readonly createSignature: (payload: string) => Promise<string>
+    private readonly useGroupKey: () => Promise<GroupKeySequence>
```

Comment:
Yes, that logic is almost followed here. But `defaultPartition` is just after `streamId` because those are about the same "topic", and helper accessor functions are after other fields.